students/anthony_mckeever/lesson5/assignment_1/mailroom.py

```diff
@@ -0,0 +1,397 @@
+    :header:        The header for the report.
+    :lengths:       The lenght of each field in the report.
+    :report_name:   The name of the report.  (Default = "Donor Report")
+    """
+    return ["\n|" + "-" * (len(sep_line) - 2) + "|",
+            f"|{report_name:^{len(sep_line) - 2}}|",
+            sep_line,
+            format_line(header, lengths, is_donor=False),
+            sep_line]
+
+
+def sort_key(item):
+    """
+    The key to sort donors by.
+    """
+    return float(item[1].replace("$", ""))
+
+
+def get_donor_summary():
+    """
+    Return a summary of all donors including their name, total donation sum, count of donations, and average donation.
+    """
+    return [(k, f"${sum(v[0]):.02f}", len(v[0]),  f"${sum(v[0]) / len(v[0]):.02f}") for k, v in donor_dict.items()]
+
+
+def format_line(item, lengths, is_donor=True):
+    """
+    Return a formatted string that will fit in the donor summary table.
+
+    :item:      The sequence of data to format.
+    :lengths:   The lengths for each field of the table.
+    :is_donor:  A boolean value determining whether or not the :item: is a donor.  (Default = True)
+    """
+    return f"| {item[0]:<{lengths[0]}} | {item[1]:>{lengths[1]}} | {item[2]:>{lengths[2]}} | {item[3]:>{lengths[3]}} |"
+
+
+def get_length(seq, name):
+    """
+    Return the max length between the longest item in a sequence or the name of the field.
+
+    :seq:   The sequence to evaluate
+    :name:  The name of the field to evaluate
+    """
+    longest = sorted(seq, key=length_key, reverse=True)[0]
+    return max(len(name), len(str(longest)))
+
+
+def length_key(item):
+    """
+    The sort key for the length of items in a sequence
+    """
+    return len(str(item))
+
+
+main_menu_dict = {"Send A Thank You":    ("\tGet prepopulated email template to thank a donor.", send_thanks),
+                  "Create a Report":     ("\t\tView a list of all donors and their cumulative donations.", create_report),
+                  "Send Letters to All": ("\tGenerate a letter for every donor.", send__to_all)
+                  }
+
+
+main_dict = {"main":   ("\tReturn to the main menu.  Can be used at any input prompt.", None),
+             "return": ("\tReturn to the main menu.  Can be used at any input prompt.", None),
+             "stop":   ("\tReturn to the main menu.  Can be used at any input prompt.", None)
+             }
+
+
+quit_dict = {"exit": ("\t\tQuit the script.  Can be used at any input prompt.", sys.exit),
+             "end":  ("\t\tQuit the script.  Can be used at any input prompt.", sys.exit),
+             "quit": ("\t\tQuit the script.  Can be used at any input prompt.", sys.exit),
+             }
+
+            
+list_dict = {"list": ("\t\tPrint a list of available donors.", print_donors),
+             "l":    ("\t\tPrint a list of available donors.", print_donors),
+             "ls":   ("\t\tPrint a list of available donors.", print_donors)
```

Comment:
it's best for above to be number options:
```
1 - send thank you
2 - create report
...
```

having the user type out the selection is error prone and not very user friendly.

Reply:
note that per directions, the program should ask to "list" or enter a donor names - user shouldn't be required to execute "list" command each time they would like to make a donation.

Reply:
> it's best for above to be number options:
> 
> ```
> 1 - send thank you
> 2 - create report
> ...
> ```
> 
> having the user type out the selection is error prone and not very user friendly.

I'll get this implemented. :)

Reply:
> note that per directions, the program should ask to "list" or enter a donor names - user shouldn't be required to execute "list" command each time they would like to make a donation.

@natasha-aleksandrova I'm a little confused by this.  I looked back through Mailroom parts 1 through 3 and the only instructions I for listing donor's is in Mailroom Part 1 under the "Send a Thank You" section which states:

> If the user types list show them a list of the donor names and re-prompt.

Mailroom Part 2 and Part 3 do not update this requirement.  I am absolutely happy to modify the code to prompt the user if they want to list before thanking the donor, but can you confirm that this is the correct and desired behavior?

Reply:
@Snip3rM00n I just saw this comment. 

The instructions are:

>  If the user types list show them a list of the donor names and re-prompt.
    If the user types a name not in the list, add that name to the data structure and use it.
    If the user types a name in the list, use it.


meaning, that those are all 3 valid options that a user can execute from "send a thank you" option. your code was requiring the user to always enter list first.

Reply:
> @Snip3rM00n I just saw this comment.
> The instructions are:
> 
> If the user types list show them a list of the donor names and re-prompt.
> If the user types a name not in the list, add that name to the data structure and use it.
> If the user types a name in the list, use it.
> 
> meaning, that those are all 3 valid options that a user can execute from "send a thank you" option. your code was requiring the user to always enter list first.

Ah, I see where the confusion is coming from.  My original code actually met these requirements, the user didn't have to enter a `list` command before being able to apply a donation to an existing donor.  The comment I had in my code on line 62:

> `# Don't include the list of donors in the help text so the user has to request them.`

was intended to mean the donor names do not appear in the list menu commands, so if they wanted to view them they could use a list command to see them.  The user could have entered `send a thank you` and then entered an existing donor like `Cresenta Starchelle` and apply a donation from that donor without listing them first.  I used this scenario for my golden path/build verification testing.

With that said, I actually really prefer the refactor makes things much more clear and removed a good amount of the complexity around it, which is always good.  Plus refactoring is always fun. 😄